### PR TITLE
Zero inflated binomial

### PR DESCRIPTION
**Pull request by fonnesbeck**
Adds zero-inflated binomial distribution, following the other zero-inflated distributions in `discrete.py`. Will eventually be superseded  by #2246 but that PR is not yet working. This will make the main zero-inflated distributions available for 3.1.

### CODE REVIEW

File: pymc3/distributions/discrete.py [hunk 1]
```diff
@@ -644,6 +644,67 @@ def _repr_latex_(self, name=None, dist=None):
...
+    ========  ==========================
+
+    Parameters
+    ----------
+    n : int
+        Number of Bernoulli trials (n >= 0).
+    p : float
+        Probability of success in each trial (0 < p < 1).
+    psi : float
+        Expected proportion of Poisson variates (0 < psi < 1)
```
Comment (aseyboldt) [inline pymc3/distributions/discrete.py]: Poisson -> Binomial

File: pymc3/distributions/discrete.py [hunk 2]
```diff
@@ -644,6 +644,67 @@ def _repr_latex_(self, name=None, dist=None):
...
+        self.psi = psi = tt.as_tensor_variable(psi)
+        self.bin = Binomial.dist(n, p)
+        self.mode = self.bin.mode
+
+    def random(self, point=None, size=None, repeat=None):
+        n, p, psi = draw_values([self.n, self.p, self.psi], point=point)
+        g = generate_samples(stats.binom.rvs, n, p,
+                             dist_shape=self.shape,
+                             size=size)
+        sampled = g * (np.random.random(np.squeeze(g.shape)) < psi)
```
Comment (aseyboldt) [inline pymc3/distributions/discrete.py:696]: What is the `squeeze` about? Doesn't this break broadcasting with `g`?

File: pymc3/distributions/discrete.py [hunk 3]
```diff
@@ -644,6 +644,67 @@ def _repr_latex_(self, name=None, dist=None):
...
+        return reshape_sampled(sampled, size, self.shape)
+
+    def logp(self, value):
+        return tt.switch(value > 0,
+                         tt.log(self.psi) + self.bin.logp(value),
+                         tt.log((1. - self.psi) + self.psi * tt.pow(1 - self.p, self.n)))
```
Comment (aseyboldt) [inline pymc3/distributions/discrete.py]: Does this work numerically for the gradient? Maybe a test that this is reasonable for corner cases for p and psi and large n would be nice.

Comment (fonnesbeck) [reply pymc3/distributions/discrete.py]: Isnt that what the tests do?

File: pymc3/distributions/discrete.py (hunk 2 above)
Comment (fonnesbeck) [reply pymc3/distributions/discrete.py:696]: Doesn't appear to. I'm just mirroring what is happening in the other ZI dists.

Comment (fonnesbeck) [reply pymc3/distributions/discrete.py:696]: It passes the `test_distribution_random` test.

File: pymc3/distributions/discrete.py (hunk 3 above)
Comment (junpenglao) [reply pymc3/distributions/discrete.py]: I think this work for the gradient but is numerically unstable. Why not use the log-sum-exp trick as in `pm.Mixture`? https://github.com/pymc-devs/pymc3/blob/master/pymc3/distributions/mixture.py#L110-L115

Comment (fonnesbeck) [reply pymc3/distributions/discrete.py]: Ah, right. I was just following the implementations of the other ZI distributions, but they should be "robustified", yes.

File: pymc3/distributions/discrete.py (hunk 2 above)
Comment (aseyboldt) [reply pymc3/distributions/discrete.py:696]: ok then

File: pymc3/distributions/discrete.py (hunk 3 above)
Comment (fonnesbeck) [reply pymc3/distributions/discrete.py]: Should be good to go now

File: pymc3/distributions/discrete.py [hunk 4]
```diff
@@ -689,9 +697,18 @@ def random(self, point=None, size=None, repeat=None):
...
     def logp(self, value):
-        return tt.switch(value > 0,
-                         tt.log(self.psi) + self.bin.logp(value),
-                         tt.log((1. - self.psi) + self.psi * tt.pow(1 - self.p, self.n)))
+        psi = self.psi
+        p = self.p
+        n = self.n
+
+        logp_val = tt.switch(value > 0,
+                 logsumexp(tt.log(psi) + self.bin.logp(value)),
```
Comment (aseyboldt) [inline pymc3/distributions/discrete.py]: Shouldn't this still be just `tt.log(self.psi) + self.bin.logp(value)`?

File: pymc3/distributions/discrete.py [hunk 5]
```diff
@@ -689,9 +697,18 @@ def random(self, point=None, size=None, repeat=None):
...
+                 logsumexp(tt.log((1. - psi) + psi * tt.pow(1 - p, n))))
```
Comment (aseyboldt) [inline pymc3/distributions/discrete.py]: I think this should be
`logaddexp(tt.log1p(-psi), tt.log(psi) + n * tt.log1p(-p))`
where
```python
def logaddexp(a, b):
    diff = b - a
    return tt.switch(diff > 0, b + tt.log1p(tt.exp(-diff)), a + tt.log1p(tt.exp(diff)))
```

File: pymc3/distributions/discrete.py [hunk 6]
```diff
@@ -606,15 +607,14 @@ class ZeroInflatedPoisson(Discrete):
...
     theta : float
         Expected number of occurrences during the given interval
         (theta >= 0).
-    psi : float
-        Expected proportion of Poisson variates (0 < psi < 1)
-
     """
 
-    def __init__(self, theta, psi, *args, **kwargs):
+    def __init__(self, psi, theta, *args, **kwargs):
```
Comment (aseyboldt) [inline pymc3/distributions/discrete.py:617]: isn't this a backward incompatible change?

**Comment by junpenglao** [comment]
Agree with @aseyboldt, the logp is off, for example:
```python
x = np.concatenate([np.random.poisson(4, size=180), np.zeros(20)])
with pm.Model() as model0:
    ψ = pm.Beta('ψ', 1., 1.)
    θ = pm.Gamma('θ', 1., 1.)
    like = ZeroInflatedPoisson('like', psi=ψ, theta=θ, observed=x)
    tr0 = pm.sample(3000, init=None, njobs=2)
pm.traceplot(tr0);
```
gives:
image

File: pymc3/distributions/discrete.py (hunk 6 above)
Comment (fonnesbeck) [reply pymc3/distributions/discrete.py:617]: This harmonizes them with the convention in `Mixture`, which is where they will eventually end up. So, a break is coming in one place or the other.

File: pymc3/distributions/discrete.py (hunk 4 above)
Comment (fonnesbeck) [reply pymc3/distributions/discrete.py]: Doesn't `logsumexp` help here?

Comment (aseyboldt) [reply pymc3/distributions/discrete.py]: The exp is missing :-)
It helps if you have an expression like log(exp(a) + exp(b)). But we are only doing log(a + b) here.

File: pymc3/distributions/discrete.py (hunk 5 above)
Comment (fonnesbeck) [reply pymc3/distributions/discrete.py]: You are correct.

Comment (fonnesbeck) [reply pymc3/distributions/discrete.py]: Actually, our `logsumexp` doesn't have the same signature, but you are right in principle.

File: pymc3/distributions/discrete.py (hunk 6 above)
Comment (aseyboldt) [reply pymc3/distributions/discrete.py:617]: Then we should definitely put that in the release notes. And maybe also print a warning until 3.2?

File: pymc3/distributions/discrete.py (hunk 5 above)
Comment (aseyboldt) [reply pymc3/distributions/discrete.py]: I think we should just add the `logaddexp` function. This is nicer (and I think faster) if we have only two variables and comes in handy quite often. I'm surprised that I can't find this in theano already....

Comment (fonnesbeck) [reply pymc3/distributions/discrete.py]: In fact, I think `logsumexp` is used incorrectly in `Mixture`, if I am reading it correctly.

Comment (junpenglao) [reply pymc3/distributions/discrete.py]: logsumexp is getting a weight vector [psi, 1.-psi], maybe that's why?

**Comment by fonnesbeck** [comment]
OK, getting something more reasonable now:

unknown

**Comment by aseyboldt** [comment]
LGTM

**Comment by junpenglao** [comment]
Tested locally, works perfect (even the case i showed you few days ago in a nb @fonnesbeck)
